plotly/tools.py

```diff
@@ -1451,6 +1451,697 @@ class FigureFactory(object):
     """
 
     @staticmethod
+    def _calc_stats(data):
+        """
+        Calculate statistics for use in violin plot.
+        """
+        import numpy as np
+
+        x = np.asarray(data, np.float)
+        vals_min = np.min(x)
+        vals_max = np.max(x)
+        q2 = np.percentile(x, 50, interpolation='linear')
+        q1 = np.percentile(x, 25, interpolation='lower')
+        q3 = np.percentile(x, 75, interpolation='higher')
+        IQR = q3 - q1
+        whisker_dist = 1.5 * IQR
+
+        # in order to prevent drawing whiskers outside the interval
+        # of data one defines the whisker positions as:
+        d1 = np.min(x[x >= (q1 - whisker_dist)])
+        d2 = np.max(x[x <= (q3 + whisker_dist)])
+        return vals_min, vals_max, q1, q2, q3, d1, d2
+
+    @staticmethod
+    def _make_half_violin(x, y, fillcolor='#1f77b4', linecolor='rgb(0,0,0)'):
+        """
+        Produces a sideways probability distribution fig violin plot.
+        """
+        from plotly.graph_objs import graph_objs
+
+        text = ['(pdf(y), y)=(' + '{:0.2f}'.format(x[i]) +
+                ', ' + '{:0.2f}'.format(y[i]) + ')'
+                for i in range(len(x))]
+        return graph_objs.Scatter(
+            x=x,
+            y=y, mode='lines',
+            name='',
+            text=text,
+            fill='tonextx',
+            fillcolor=fillcolor,
+            line=graph_objs.Line(width=0.5, color=linecolor, shape='spline'),
+            hoverinfo='text',
+            opacity=0.5
+        )
+
+    @staticmethod
+    def _make_rugplot(vals, pdf_max, distance, color='#1f77b4'):
```

Comment:
🐄 ooh, ya, this is the same color as in https://github.com/plotly/plotly.py/pull/471/files#r65923768! Give it a nice, clear constant :)